### DISCUSSION TITLE
Advanced dynamic liquidation feature

### DISCUSSION THREAD

**Discussion by bonomat** [comit-network/baru#29] [Ideas]
PR #25 introduces a new liquidation branch to the lending contract which allows the lender to take all the collateral if the bitcoin price falls below a certain value. For that he needs to provide the signature and values from oracle. 

The downside of this is that the borrower will lose all the collateral if the price tanks. 

Ideally the protocol would behave as following: 

- Anyone can liquidate the loan contract. _Liquidating_ is very similar to _paying back_ the loan, i.e. the lender receives back his initial investment + interest. 
- The difference is that anyone can do this. In order to incentivize others to payback the loan, the bitcoin in the contract should be offered slightly below market value (e.g. -5%). 
- Only as much collateral should be sold to the liquidator to pay back the loan. 
- The borrower should still receive back the remaining bitcoin. 

For example: 
- current rate for `1 BTC` is `$100` 
- The borrower takes a loan of $100 and puts up 2BTC as collateral (LTV ratio is at 50% = `principal_amount/collateral => 100/200 => 0.5 `). 
- They agree to liquidate the loan if the LTV ratio goes above 80%
- BTC price crashes to `$60` (LTV ratio is 83% `100/2*60 = 0.833`)
-  A liquidator can now pay back $100 and receives $100 worth of Bitcoin at the rate of $60-5%, e.g. 1.66 BTC. 
- The lender receive `$100` (we ignoring any interest rates)
- The borrower receives `0.33` BTC (`2 - 1.66`)

_please excuse any calculation errors_

This reduces the risk for the borrower to lose all his Bitcoin _just_ because the price tanked. 
Everyone in this scenario should be happy: 
- the lender received his funds back
- the liquidator got a good deal on the bitcoin
- the borrower _only_ lost 5%.

**Comment by thomaseizinger** [comment]
I think this actually describes two features:

1. Liquidation by any party, not just the lender.
2. Partial liquidation, liquidate prior to the point where the entire collateral is worth the amount of the loan.

Is my understanding correct?

**Reply by bonomat** [reply]
Yes, I think so. 
Do you see a way to consider these in separation? 
I think these two are tightly coupled:

Liquidation in this case only makes sense if we talk about the above mentioned approach of liquidation where the lender gets compensated. It wouldn't work if the liquidator would get all the collateral.

This means, if we allow anyone to liquidate the loan we need to consider two things: 
1. the lender should get compensated: i.e. get back his money + interest
2. the liquidator should get some of the collateral but has to pay back the debt
3. the borrower should not lose his full collateral

I think the borrower would have a big disadvantage if we have 2. without 3.

**Reply by thomaseizinger** [reply]
See https://github.com/comit-network/baru/discussions/29#discussioncomment-1084941 but in short:

If we simply set the terms of the loan such that you can only liquidate when the collateral is worth exactly the borrowed amount, (3) should not be an issue.

**Reply by bonomat** [reply]
If the lender then receives the collateral asset he risks on losing money. 
He will need to sell that asset to cover his loss, if the price of the asset keeps tanking, he might not find someone in time.

**Reply by thomaseizinger** [reply]
>  He will need to sell that asset to cover his loss, if the price of the asset keeps tanking, he might not find someone in time.

Isn't that an inherent issue that you need to accept when using a volatile asset as a collateral?

**Reply by bonomat** [reply]
That's why we want to countersteer with over-collateralization and liquidate before the value dropps too far.

**Reply by thomaseizinger** [reply]
That makes sense. I still don't understand though why we can't consider these things separate.

If we have a collateral of X BTC worth $150 for a loan of $100, we can open up a liquidation path at the price of $110 for the entire amount of X to repay the loan. This offers the BTC at a discount to the market.

We can build that so anyone can trigger this and it is still fair to the borrower because the value of his collateral is so close to the actual value that loosing the additional $10 can be argued is just the risk of not putting enough collateral down.

**Reply by bonomat** [reply]
Ah. That is a possibility. 

Imho it is not favorable for the borrower though. In this scenario: 
- the lender profits because he receives what he wanted (the amount he lent out + interest)
- the liquidator gets a great deal on BTC
- the borrower loses the entire collateral. 

We could argue that this is the risk of taking such a loan but from the borrower perspective I think this is very risky. Ideally we would at least allow him to add more collateral instead.

**Reply by thomaseizinger** [reply]
> We could argue that this is the risk of taking such a loan but from the borrower perspective I think this is very risky. Ideally we would at least allow him to add more collateral instead.

In a UTXO-model, adding more collateral means we actually spend and invalidate the old contract, right? That sounds like a completely different feature altogether, no?

> the borrower loses the entire collateral.

Well yes, "entire" because it is gone to shit :sweat_smile: 
The assumption is that they don't actually want it back and rather keep the borrowed money because it is worth more than their collateral.

**Reply by DeliciousHair** [reply]
I don't think this option makes complete sense though.

The LTV essentially sets the risk level for both parties, so if there is an agreed upon LTV that's it. What would agreed-upon + a bit more even be other than tilting the scale towards the lender by some amount (and what amount would that even be?) once an agreement was made?

For example, let's assume the lender creates a loan with a high interest rate and low (near unity) LTV. They would basically be betting against the borrower that the price of BTC is going to remain basically flat and is using this as a play to get a bargain on BTC by way of the fact they want to the price of BTC to be low enough that it makes sense for the borrower to default.

If the borrower had to put up a bit of extra collateral though, this would lower the LTV and potentially destroy the lenders ability to win their own bet.

**Reply by bonomat** [reply]
> In a UTXO-model, adding more collateral means we actually spend and invalidate the old contract, right? That sounds like a completely different feature altogether, no?

Yes, that's a completely different feature. That's something we can add in the next iteration :D 

> Well yes, "entire" because it is gone to shit 😅
> The assumption is that they don't actually want it back and rather keep the borrowed money because it is worth more than their collateral.

That assumption is only true if the price crashes below the loan value. 
Does my LTV example above make sense to you? 

With an incentive model where everyone can liquidate we can protect the borrower from bigger losses.  


@DeliciousHair : 

> If the borrower had to put up a bit of extra collateral though, this would lower the LTV and potentially destroy the lenders ability to win their own bet.

This sounds like an issue if the lender's business model is a "bet" and not to make money through interest rate? 

The higher the collateral (and the lower the LTV ratio), the more likely it is that the borrower will pay back the loan including interest. Which means, it would be in favor of the lender if the borrower puts up more collateral.

**Comment by luckysori** [comment]
I'm struggling with the terminology here: I (we?) have been calling _liquidation_ the act of taking the collateral and giving it to the lender (for now in full). Here the loan is being _repaid_, albeit by another party.

Browsing the internet it seems like I may have been using _liquidation_ wrong. Should we use _default_ instead of _liquidation_? Any other ideas?

**Reply by bonomat** [reply]
Sorry for the confusion. I'm also still learning and to my understanding, _traditional_ finance is also not perfectly clear on this. 

My understanding is: 
- Defaulting is when a borrower falls behind on his payments, i.e. he missed a payment
- Liquidation is the process of selling assets, i.e. you make them liquid
- Repayment is the process of settling the deb, i.e. you repay the loan principal amount.

**Reply by thomaseizinger** [reply]
Here is my take: We are _liquidating_ the collateral to _repay_ the loan.

This can happen for two reasons:

1. Prior to the end of the loan term, the collateral is no longer worth as much as the borrowed amount. The collateral needs to be liquidated to ensure that the lender is being repaid at all as there is no more incentive for the borrower to do so.
2. At the end of the loan term, the borrower could _default_ by failing to repay the loan (for whatever reason). In such a case, it needs to be possible for the lender (or potentially anyone) to acquire enough of the collateral to repay the loan.

**Reply by luckysori** [reply]
Can you help us, @DeliciousHair?

**Reply by DeliciousHair** [reply]
All of the above sounds correct to me, but @thomaseizinger is, I think, mixing the two things together is all--which I don't think is appropriate here.

The subtle but important issue here is that because this is all BTC-based, there is no way to set the LTV low enough to make the proposition risk-free for the lender owing to the volatility of BTC; the borrower would need to put up infinite collateral for that. So there is the price-base liquidation trigger that the lender can use once the current BTC/USD (assuming USD is the trading pair) price falls below the point where collateral * exchange_rate <= loan + interest, in which case the lender gets all the collateral as payment in full and closes the contract, possibly taking a loss.

On the other hand, defaulting means that the borrower has simply not repaid the loan / missed a payment for whatever reason. As the current model only works on a single payment approach, the liquidation mechanism is able to be triggered here not because of a price signal but a temporal one, and the assets are split up--the lender only gets what they are entitled to (principal + interest), and any remaining collateral goes back to the borrower.

**Comment by luckysori** [comment]
With [this](https://github.com/comit-network/baru/discussions/29#discussioncomment-1084901) comment in mind and given the title of this discussion, I was expecting this feature proposal to look something like:

1. Collateral value drops below threshold.
2. Lender is offline, but a third party comes along and spends the collateral contract by:
    - Giving the lender the collateral minus a fee.
    - Giving themself a predetermined fee.
    
This version still sees the lender possibly taking more collateral than they deserve (it should just be capped at the repayment amount), but that is already the case for all forms of ~liquidation~ defaulting that currently exist. I would see fixing that as a separate feature.

**Reply by thomaseizinger** [reply]
It is my understanding as well that we can first introduce a feature where liquidation is triggered by a third-party.

How much of the collateral is being paid back should be orthogonal to this. If we don't want the borrower to loose all the collateral, we should simply set the terms of the loan in such a way that _at the time of liquidation_, the collateral is exactly worth the borrowed amount (+ some fees for incentives).

**Reply by bonomat** [reply]
> Lender is offline, but a third party comes along and spends the collateral contract by:
> 
>     * Giving the lender the collateral minus a fee.


The goal should be that the lender receives his initial asset type back. I understand that he _might_ accept the collateral asset type but to cover his loss he would need to sell it to receive his initial asset back.

**Reply by luckysori** [reply]
> The goal should be that the lender receives his initial asset type back

That's a good point. If the price has dropped below the threshold, it is convenient for the lender to get a repayment instead of taking control of the collateral. Taking the collateral is a last resort. Correct?

**Reply by bonomat** [reply]
Yes, I would say so.

**Comment by luckysori** [comment]
And now onto the actual feature that Philipp mentions here :stuck_out_tongue:

I think it's weird to allow anyone else to repay the loan, because you would be preventing the borrower from benefiting from the short position on the collateral asset which they took when they accepted the loan.

If someone else can come along and take the cheap collateral that you bet you would be able to repurchase that kind of defeats the purpose of this product (for the borrower).

**Reply by bonomat** [reply]
> If someone else can come along and take the cheap collateral that you bet you would be able to repurchase that kind of defeats the purpose of this product.

That's actually a feature which is common in today's defi world. For example MakerDAO ensures the stability of their sable coin using this approach https://github.com/makerdao/community/blob/master/faqs/liquidation.md

If the lender wants the _cheap_ collateral he can liquidate himself as well of course.

**Reply by luckysori** [reply]
Okay, so it's only open to global repayment when it drops below a threshold. And the borrower still has exclusive access to the repayment if the price has not dropped below said threshold yet. Am I understanding it correctly?

**Reply by bonomat** [reply]
Mhmm.. I did not mention the _exclusive access_ . I guess that's another feature you could build on top? I would have anyone allowed to repay if the TLV ratio is in a critical range.

**Reply by luckysori** [reply]
I think we're talking about the same thing:

- If the value of the collateral is still in good standing, only the borrower can repay. The price may have moved in his favour, meaning that he gets to repurchase cheaper collateral asset, realising his short position.
- If the value of the collateral drops sufficiently, anyone can repay. It's worse for the borrower if someone other than him repays, but the argument is that he's had the opportunity to do so until the value of the collateral dropped further.

**Comment by thomaseizinger** [comment]
I think I understand better now, which different features could be added here. I'll try a summary of things I see as orthogonal:

1. Liquidation of the collateral should be tied to repayment of the loan:

The lender should receive their original asset back instead of the collateral. Hence, the loan contract should enforce that a liquidation branch can only be triggered if someone repays the loan in the original currency to the borrower. This someone is likely a different party that is interested in acquiring the collateral. The contract would only enforce the presence of the repayment output, (I think) what happens to the collateral doesn't matter at this stage but obviously, whoever pays back the loan will transfer the collateral to themselves. To incentivize liquidation, it can be triggered already at a price of say 110% of the original loan.

2. Adding collateral to the loan to avoid liquidation:

To avoid the situation of described in (1), a borrower should be able to take the loan contract and add more collateral to it. I am not sure if this is doable in a non-interactive manner. We would have to enforce presence of the loan contract with different terms in the output again. But to do that non-interactively, we would already need to know the new terms of the loan at the time of creating it. Maybe this can be solved by upping the collateral in fixed intervals like 10% or 20%?

Alternatively, this branch could require interactivity with the borrower to co-sign the new terms.

3. Liquidation of the collateral way before its value approaches the lent amount

I see this as a special case of (1) because it basically makes the mentioned 10% configurable and gives some of the collateral back to the borrower to make it fairer. In particular, if we define liquidation to be valid at 130% of loan value, we can allow repayment by claiming the equivalent of 110% and give the remaining BTC back to the borrower.

**Reply by DeliciousHair** [reply]
@thomaseizinger while I get what you're saying in terms of how the transaction would actually be put together, I don't see this as a good option (unless I'm misreading this, which is quite possible) as it would lead to a situation where third-parties could game the system in a risk-free manner, thus putting all other parties at risk for no benefit that I can see.

For example, what if some loan has an LTV of something like 0.1, but a liquidation trigger that corresponds to an LTV of, say, 0.5? In this case, the third party can swoop in and collect all the collateral for what amounts to a bargain-basement price, and the borrower is left holding the bag.

It's for this reason I do not see point 3 as needing to exist, as the lender needs to assume some risk in this transaction. ie, the LTV defines the price-signal for liquidation.

Finally, your point 2 I 100% agree with, as it is effectively the same thing as the contract having a mechanism analogous to a margin call in place. This makes perfect sense for this product, given the need for the whole thing to be collateralized with BTC.
